lang/shared/src/main/scala/com/wavesplatform/lang/TypeChecker.scala

```diff
@@ -5,28 +5,39 @@ import cats.syntax.all._
 import com.wavesplatform.lang.Terms._
 import com.wavesplatform.lang.ctx.{Context, PredefType}
 import monix.eval.Coeval
+import ctx._
 
 import scala.util.{Failure, Success, Try}
 
 object TypeChecker {
 
   type TypeDefs     = Map[String, TYPE]
-  type FunctionSigs = Map[String, (List[TYPE], TYPE)]
+  type FunctionSigs = Map[String, FUNCTION]
   case class TypeCheckerContext(predefTypes: Map[String, PredefType], varDefs: TypeDefs, functionDefs: FunctionSigs)
 
   object TypeCheckerContext {
     val empty = TypeCheckerContext(Map.empty, Map.empty, Map.empty)
 
     def fromContext(ctx: Context): TypeCheckerContext =
-      TypeCheckerContext(predefTypes = ctx.typeDefs,
-                         varDefs = ctx.letDefs.mapValues(_.tpe),
-                         functionDefs = ctx.functions.mapValues(_.signature))
+      TypeCheckerContext(
+        predefTypes = ctx.typeDefs,
+        varDefs = ctx.letDefs.mapValues(_.tpe),
+        functionDefs = ctx.functions.mapValues(x => FUNCTION(x.args.map(y => y._2), x.resultType))
+      )
   }
 
   type TypeResolutionError      = String
   type TypeCheckResult[T]       = Either[TypeResolutionError, T]
   private type SetTypeResult[T] = EitherT[Coeval, String, T]
 
+  private def resolveTypes(typeWithParam: TYPE, map: Map[String, Either[String, TYPE]]): Either[String, TYPE] =
```

Comment:
are you sure `Map[String, Either[String, TYPE]])` is neccessary and `Map[String,TYPE]` won't be enough?